### PR TITLE
Use time zone when casting date to timestamp.

### LINKED ISSUE #2773: More time zone issues
The SQL logic tests fail when run after 8pm in New York. The failing test is
```
query T
SELECT now() - current_date()::timestamp < interval '24h10s'
----
true
```

Expanding those two expressions, we see that `now()` is `2015-10-06 01:42:42.244354302 +0000 UTC` and `current_date()::timestamp` is `2015-10-05 00:00:00 +0000 UTC`. `current_date()` is operating in the server's time zone (EDT) instead of the session time zone (UTC). 

`current_date()` is currently a separate code path from `current_timestamp()::date`. We need to either provide some way to mock out the EvalContext timestamps for testing or redefine `current_date()` in terms of a cast so it can't diverge in this way.

There's also an asymmetry in the cast operators: casting a timestamp to a date considers the session time zone, but casting a date to a timestamp does not. This seems incorrect.

### PR DESCRIPTION
**Pull request by vivekmenezes**
Fix current_date.

closes #2773

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
i might not have a chance to review this today; could you hold off on merging until i have?

**Comment by vivekmenezes** [comment]
@tamird let's wrap this up soon because the tests fail every night without this change.

File: sql/parser/datum.go [hunk 1]
```diff
@@ -435,9 +429,8 @@ func (d DTimestamp) IsMin() bool {
 	return d.Before(d.Add(-1))
 }
 
-// TODO:(vivek) implement SET TIME ZONE to improve presentation.
 func (d DTimestamp) String() string {
```
Comment (tamird) [inline sql/parser/datum.go]: this is really only used in parser tests, right?

File: sql/parser/eval.go [hunk 2]
```diff
@@ -1246,8 +1257,7 @@ func ParseDate(s DString) (DDate, error) {
 	if err != nil {
 		return DummyDate, err
 	}
-
-	return MakeDDate(t), nil
+	return DDate{Time: t}, nil
```
Comment (tamird) [inline sql/parser/eval.go]: don't we need to force this to UTC?

**Comment by tamird** [comment]
I'm uneasy about this change because of the inherent subtlety. Can we jump straight to changing Date to a triplet of year,month,day?

File: sql/parser/datum.go (hunk 1 above)
Comment (vivekmenezes) [inline sql/parser/datum.go]: Only in eval_test.go

File: sql/parser/eval.go (hunk 2 above)
Comment (vivekmenezes) [inline sql/parser/eval.go]: The time.Parse() documentation says, "In the absence of a time zone indicator, Parse returns a time in UTC."

**Comment by vivekmenezes** [comment]
I agree that change is needed here. Filed away https://github.com/cockroachdb/cockroach/issues/2803

**Comment by vivekmenezes** [comment]
Can I merge this?

**Comment by tamird** [comment]
LGTM

**Pull request merged**